src/testing/conditional_independence.py

```diff
@@ -0,0 +1,159 @@
+# -*- coding: utf-8 -*-
+"""
+Functions for performing permutation-based conditional independence tests.
+"""
+import numpy as np
+
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import r2_score
+
+import seaborn as sbn
+import matplotlib.pyplot as plt
+
+from tqdm import tqdm
+
+
+def _check_array_lengths(array_1, array_2, array_3):
+    if (array_1.size != array_2.size) or (array_1.size != array_3.size):
+        msg = 'All arrays MUST be of the same size.'
+        raise ValueError(msg)
+    return
+
+
+def _ensure_is_array(array_arg, name):
+    if not isinstance(array_arg, np.ndarray):
+        msg = '{} MUST be an instance of np.ndarray.'.format(name)
+        raise TypeError(msg)
+    return
+
+
+def _make_regressor(x_2d, y):
+    # Note, linear regression used instead of Random Forest Regression since
+    # the conditional independence test based on RandomForest regression did
+    # not pass the unit tests meant to assess whether the test statistic was
+    # a u-statistic.
+    regressor = LinearRegression()
+    regressor.fit(x_2d, y)
+    return regressor
+
+
+def computed_vs_obs_r2(x1_array,
+                       x2_array,
+                       z_array,
```

Comment:
Should make the z_array optional so this function can be used for both marginal and conditional independence tests.